Simplify config generation in the tests

AndrejMitrovic opened the pull request:
It's not the most efficient code because of the `.array` calls, but it should not matter for the vast majority of tests as this is only the set-up stage. And I wanted to keep the code reasonably clean.

AndrejMitrovic committed:
Tests: simplifly the node config creation

The logic became too hard to follow.

AndrejMitrovic committed:
Make file and line runtime args

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1335,8 +1335,8 @@ public struct TestConf
-public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager, string file = __FILE__, int line = __LINE__)(
-    in TestConf test_conf)
+public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager,
+    string file = __FILE__, int line = __LINE__)(in TestConf test_conf)
```
I'd rather we move them to runtime arguments, what do you think ?

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1389,117 +1376,111 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager,
 
-    Config makeMainConfig (size_t idx, NodeConfig self, NodeConfig[] node_confs,
-        ValidatorConfig validator_self = ValidatorConfig.init)
+    immutable(Address[]) makeNetwork (size_t idx, Address[] addresses)
```
The name of this function is not very accurate

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1389,117 +1376,111 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager,
-                .filter!(conf => conf != self)
-                .map!(conf => conf.address);
+        if (!test_conf.configure_network ||
+            addresses.length == 0)  // One NetworkManager test uses 1 validator
```
Can we just fix that test ?

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1389,117 +1376,111 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager,
-    extra_node_configs.enumerate(test_conf.outsider_validators).each!(
-        pair => main_configs ~= makeMainConfig(
-            pair.index % node_configs.length, pair.value, node_configs));
+    alias validatorAddress = (idx, key) => format("Validator #%s (%s)", idx,
+        key.address);
+
+    alias fullNodeAddress = (idx) => format("FullNode #%s", idx);
```
Ergh, to me it's clearer to have it as a nested function. This is a template literal, but it doesn't need to be templated.

AndrejMitrovic replied:
okay _sad noises_

Geod24 reviewed (changes requested):
Both versions don't seem very clear to me TBH. Made a few comments.

AndrejMitrovic committed:
Fix validator counts in the tests

Many tests used between 1 - 3 validator nodes,
but they should use ideally 4 nodes.

Added a check in makeTestNetwork() to force
spawning at least 4 validator nodes.

Some tests spawn many more nodes so we cannot
hardcode it to the validator set in the
non-generated TestGenesisBlock until those
tests are fixed.

AndrejMitrovic commented:
It turns out a lot of the tests were assuming < 4 validator nodes. I fixed them. There are still tests which spawn many enrolled validators, but those can be fixed later.

AndrejMitrovic commented:
It's green now.

Geod24 commented:
Noice
